File: src/main/java/ting/leetcode/java/binarytree/BinaryTreeInorderTraversal.java
package ting.leetcode.java.binarytree;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BinaryTreeInorderTraversal {
    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null) {
            return Collections.emptyList();
        }
        ArrayList<Integer> record = new ArrayList<>();
        record.addAll(inorderTraversal(root.left));
        record.add(root.val);
        record.addAll(inorderTraversal(root.right));
        return record;
    }
}

File: src/main/kotlin/ting/leetcode/tree/BinaryTreeInorderTraversal.kt
package ting.leetcode.tree

class BinaryTreeInorderTraversal {
    fun inorderTraversal(root: TreeNode?): List<Int> {
        if (root == null) return emptyList()

        val result = mutableListOf<Int>()
        val stack = ArrayDeque<TreeNode?>()
        var currentNode = root;

        while (currentNode != null || stack.isNotEmpty()) {
            stack.run {
                if (currentNode != null) {
                    addFirst(currentNode)
                    currentNode = currentNode?.left
                } else {
                    currentNode = removeFirst()
                    result.add(currentNode?.`val`!!)
                    currentNode = currentNode?.right
                }
            }
        }
        return result
    }

    fun inorderTraversalRecursion(root: TreeNode?): List<Int> =
        if (root == null)
            emptyList()
        else
            mutableListOf<Int>().apply {
                addAll(inorderTraversalRecursion(root.left))
                add(root.`val`)
                addAll(inorderTraversalRecursion(root.right))
            }
}
